text/0000-feature-metadata.md

```diff
@@ -0,0 +1,275 @@
+    with `since` and/or `note` keys. Cargo will warn downstream crates using this
+-   `public`: A boolean flag defaulting to `true` that indicates whether or not
+    downstream crates should be allowed to use this feature
+-   `unstable`: A boolean flag indicating that the feature is only usable with
+    nightly Rust, or is otherwise not API-stable
+
+If a downstream crate attempts to use the features `baz` and `qux`, they will
+see messages like the following:
+
+```
+warning: feature `quux` on crate `mycrate` is deprecated since version
+  1.2.3: "don't use this!"
+
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Validation and parsing of the new schema, described above, shoud be relatively
+straightforward. Each of the added keys is discussed individually in the
+following sections:
+
+## `requires`
+
+`requires` will take the place of the feature dependency array that currently
+exists. Semantics will remain unchanged.
+
+## `doc`
+
+`doc` is the most straightforward: it accepts markdown-flavored text, and should
+be thought of as the equivalent to a `#[doc(...)]` attribute. Like doc comments,
+the first line should be usable as a summary, though most use cases will only
+have one line.
+
+There is nothing in this RFC that cargo `must` do with this action, since it is
+mainly intended for the consumption of `rustdoc` or `docs.rs`. However, it can
+be used for general diagnostic information.
+
+## `deprecated`
+
+`deprecated` should be thought of as the equivalent of the [`deprecated`]
+attribute in Rust source. The value can be a boolean, string, or an object with
+`since` or `note` keys. Schema rules are as follows:
+
+-   If a boolean value, `false` indicates not deprecated and `true` indicates
+    deprecated
+-   If an object, the keys `since` and/or `note` can be specified
+-   An empty object is not allowed to avoid the ambiguous
+    `foo = { deprecated = {} }`
+-   A string `foo = { deprecated = "my msg" }` will be equivalent to if that string
+    was specified in the `note` field:
+    `foo = { deprecated = { note = "my msg" } }`
+-   If not specified, the default is `false`
+
+If a downstream crate attempts to use a feature marked `deprecated`, Cargo
+should produce a warning.
+
+## `public`
+
+`public` is a boolean value that defaults to `true`. It can be thought of as
+`pub` in Rust source files, with the exception of being true by default. If set
+to `false`, Cargo should forbid its use with an error message on any downstream
+crates.
+
+There needs to be an escape hatch for this for things like benchmarks - RFC TBD
+on how this works.
+
+## `unstable`
+
+`unstable` is a boolean value that defaults to `false`. It should indicate that
+a feature gates something that is not API-stable, or can only be built using
+`nightly` Rust.
```

Comment:
Is this purely informational or should there be some semantics to it?

Personally I suspect there is enough for us to figure out about this one in further discussions that this should be postponed, rather than being included in the RFC so as to not slow down the approval of the rest of this.

One idea for how we could act on `unstable` if we move forward with it in this RFC:
- `unstable` features may only be required by other unstable features or by the command-line
  - Since `unstable` features are not subject to semver, they may be removed, so any non-existent features that an unstable feature requires produces a warning at most **unless** the unstable feature is activated, then it errors (no idea how complicated this would be)
- Future possibility: `unstable` features may also be required when the version requirement is `=<x>.<y>.<z>`

Reply:
I am also a bit unsure about this one and don't mind dropping it. @ehuss do you have any comments since you wrote https://github.com/rust-lang/cargo/issues/10881?

If there is a way to use features marked non-public (https://github.com/rust-lang/rfcs/pull/3416#discussion_r1172889980) then that solves the use case I had (using [`visibility`](https://crates.io/crates/visibility) to expose internal functions & types for benchmarking / integration tests)

Reply:
I have removed unstable from this RFC in f1fb220ddff01ec8835cda06199ff29605f8489f